doc/development/guide/documentation.rst

```diff
@@ -1,39 +1,455 @@
+is apparent from the signature, it is sufficient for the docstring to simply consist of the single
+line summary sentence.
+
+Some notes on the above structure:
+
+* **Function summary:** The summary should provide a brief, basic description of the function. Do not
+  include the function or argument names in the summary. It is important that the initial summary
+  be a single sentence; Sphinx will truncate the summary at the first period otherwise.
+
+* **Arguments:** To describe named function arguments in the signature. If the type is a PennyLane class or
+  function, use the syntax ``(~.Operation)``; do **not** use a Sphinx role such as ``:class:`` or ``:func:``.
+  The level of type description is up to the discretion of the code author and reviewers; acceptable
+  examples include ``(array)``, ``(array[float])``, ``(dict)``, ``(dict[str, int])``, ``(Sequence[int] or int)``.
+  Note that square brakets are used to define the required types of container elements.
+
+  *Note:* if the argument has a default value, *do not* include it in the description if it is declared
+  in the signature. Sphinx will automatically extract the default and render it with the argument
+  description.
+
+* **Keyword Arguments:** To describe variable length named keyword arguments provided via
+  ``**kwargs``. Provide the default values if relevant.
+
+* **Returns:/Yields:** To describe the return (yielded) value of the function (generator). This may
+  be omitted if the function signature/summary already sufficiently describe both the return value
+  and type.
+
+* **See also:** Provide optional links to relevant pages, functions, classes, methods, etc.
+
+* **Example:** To provide a minimal working example showing basic usage of the function. The example
+  should be *minimal* (reduce line counts where possible), but complete; a reader should be able to
+  copy and paste the example and get the same output. See :ref:`code_examples` for guidelines writing
+  useful code examples in docstrings.
+
+* **Usage details:** To provide a more complicated usage details showing different edge cases and
+  advanced usage of the function, as well as implementation details. This section of the docstring
+  is *collapsed by default* in the rendered Sphinx documentation to avoid overwhelming the reader with
+  information, however will be displayed in full to users in the terminal/Jupyter notebooks, so
+  do not let this section become too long.
+
+
+Classes
+~~~~~~~
+
+The class docstring is placed directly below the class definition:
+
+.. code-block:: python
+
+    class MyClass:
+        """Single sentence that summarizes the class.
+
+        Multi-line description of the class (optional, if required).
+
+        Args:
+            arg1 (type): Description.
+                Continuation line is indented if needed.
+            arg2 (type): description
+
+        Keyword Args:
+            kwarg1 (type): description
```

Comment:
I'm not sure I follow which one you are referring to @thisac?

Reply:
Just a nitpick; I meant that here it says `kwarg1 (type): description` while a few lines above, in functions and methods, it says `kwarg1 (type): Description. Provide the default value if applicable.`. I understand that both types are accepted (since it's described a bit further down under Punctuation and spelling), although when reading I found it a tiny bit confusing regarding what convention was to be used. Could perhaps the Punctuation and spelling section be moved up so that the reader would know those rules before being presented with these examples?

Reply:
Good point, I'll move punctuation and spelling up!

Reply:
Done